lib/plugins/cors.js

```diff
@@ -73,66 +120,234 @@ function matchOrigin(req, origins) {
+    assert.object(options, 'options');
+    assert.optionalBool(options.credentials, 'options.credentials');
+    assert.optionalArrayOfString(options.exposeHeaders,
+                                 'options.exposeHeaders');
+    assert.optionalArrayOfString(options.allowHeaders, 'options.allowHeaders');
+    assert.optionalNumber(options.preflightMaxAge, 'options.preflightMaxAge');
 
-    var headers = (opts.headers || []).slice(0);
-    var origins = opts.origins || ['*'];
+    var opts = options || {};
+    var credentials = opts.credentials || false;
+    // convert all origin values into regexps. however, leave '*' as is for
+    // easier grokking later.
+    var origins = (opts.origins || ['*']).map(toRegExp);
+    var allowHeaders = opts.allowHeaders || [];
+    var exposeHeaders = opts.exposeHeaders || [];
+    var preflightMaxAge = opts.preflightMaxAge || AC_MAX_AGE_MS;
+    var preflightStrategy = opts.preflightStrategy;
+
+    // assert origin patterns, everything should be a regexp unless it's a *
+    assert.array(origins, 'origins');
+    origins.forEach(function (o) {
+        assert.ok(o === '*' || o instanceof RegExp, o +
+                  ' is not a valid origin');
+    });
 
+    // concat customized headers to the default list
     EXPOSE_HEADERS.forEach(function (h) {
-        if (headers.indexOf(h) === -1) {
-            headers.push(h);
+        if (exposeHeaders.indexOf(h) === -1) {
+            exposeHeaders.push(h);
+        }
+    });
+
+    ALLOW_HEADERS.forEach(function (h) {
+        if (allowHeaders.indexOf(h) === -1) {
+            allowHeaders.push(h);
         }
     });
 
-    // Handler for simple requests
-    function restifyCORSSimple(req, res, next) {
-        var origin;
 
-        if (!(origin = matchOrigin(req, origins))) {
-            next();
-            return;
+    // use the preflight spec from w3
+    // http://www.w3.org/TR/cors/#access-control-request-headers-request-header
+    function preflight(req, res, next) {
+
+        // 1) if not an options request, move on.
+        if (req.method !== 'OPTIONS') {
+            return next();
         }
 
-        function corsOnHeader() {
-            origin = req.headers.origin;
+        // now check the preflight requirements are all available. if anything
+        // is missing, or the resource requested for is missing, returna 400.
+        var reqOrigin = req.headers.origin;
+        var reqMethod = req.headers[AC_REQ_METHOD];
+        var reqHeaders = req.headers[AC_REQ_HEADERS];
+        var errMsg = 'invalid preflight request';
+        var errMsg2 = 'unsupported method';
 
-            if (opts.credentials) {
-                res.setHeader(AC_ALLOW_ORIGIN, origin);
-                res.setHeader(AC_ALLOW_CREDS, 'true');
-            } else {
-                res.setHeader(AC_ALLOW_ORIGIN, origin);
+        // 2) check that the incoming preflight headers are ok
+        if (!reqOrigin || !reqMethod || !reqHeaders) {
```

Comment:
Love pulling the values up off of the `req.headers` object, this reads really well :heart: